OST2.design/tools/perl-5.8.0/man/man3/Tk.3
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "TK 1"
.TH TK 1 "2000-12-30" "Tk800.023" "perl/Tk Documentation"
.SH "NAME"
Tk \- a graphical user interface toolkit for Perl
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 3
\&    use Tk;
\&    $top = new MainWindow;
\&    MainLoop;
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The Perl/Tk manual is split up into a number of sections:
.Sh "Introduction"
.IX Subsection "Introduction"
.IP "\(bu" 4
Tk::overview
.IP "\(bu" 4
Tk::UserGuide
.Sh "Tk Geometry Management"
.IX Subsection "Tk Geometry Management"
.IP "\(bu" 4
Tk::Adjuster
.IP "\(bu" 4
Tk::form
.IP "\(bu" 4
Tk::grid
.IP "\(bu" 4
Tk::pack
.IP "\(bu" 4
Tk::place
.IP "\(bu" 4
Tk::Table
.IP "\(bu" 4
Tk::Tiler
.IP "\(bu" 4
Tk::Wm
.Sh "Binding Events and Callbacks"
.IX Subsection "Binding Events and Callbacks"
.IP "\(bu" 4
Tk::After
.IP "\(bu" 4
Tk::bind
.IP "\(bu" 4
Tk::bindtags
.IP "\(bu" 4
Tk::callbacks
.IP "\(bu" 4
Tk::Error
.IP "\(bu" 4
Tk::event
.IP "\(bu" 4
Tk::exit
.IP "\(bu" 4
Tk::fileevent
.IP "\(bu" 4
Tk::IO
.Sh "Tk Image Classes"
.IX Subsection "Tk Image Classes"
.IP "\(bu" 4
Tk::Animation
.IP "\(bu" 4
Tk::Bitmap
.IP "\(bu" 4
Tk::Compound
.IP "\(bu" 4
Tk::Image
.IP "\(bu" 4
Tk::Photo
.IP "\(bu" 4
Tk::Pixmap
.Sh "Tix Extensions"
.IX Subsection "Tix Extensions"
.IP "\(bu" 4
Tk::Balloon
.IP "\(bu" 4
Tk::BrowseEntry
.IP "\(bu" 4
Tk::DialogBox
.IP "\(bu" 4
Tk::DirTree
.IP "\(bu" 4
Tk::DItem
.IP "\(bu" 4
Tk::InputO
.IP "\(bu" 4
Tk::LabFrame
.IP "\(bu" 4
Tk::Mwm
.IP "\(bu" 4
Tk::NoteBook
.IP "\(bu" 4
Tk::TixGrid
.IP "\(bu" 4
Tk::tixWm
.IP "\(bu" 4
Tk::TList
.IP "\(bu" 4
Tk::Tree
.Sh "Tk Widget Classes"
.IX Subsection "Tk Widget Classes"
.IP "\(bu" 4
Tk::Button
.IP "\(bu" 4
Tk::Canvas
.IP "\(bu" 4
Tk::Checkbutton
.IP "\(bu" 4
Tk::Entry
.IP "\(bu" 4
Tk::Frame
.IP "\(bu" 4
Tk::HList
.IP "\(bu" 4
Tk::Label
.IP "\(bu" 4
Tk::Listbox
.IP "\(bu" 4
Tk::Menu
.IP "\(bu" 4
Tk::Menubutton
.IP "\(bu" 4
Tk::Message
.IP "\(bu" 4
Tk::Optionmenu
.IP "\(bu" 4
Tk::Radiobutton
.IP "\(bu" 4
Tk::Scale
.IP "\(bu" 4
Tk::Scrollbar
.IP "\(bu" 4
Tk::Text
.IP "\(bu" 4
Tk::Toplevel
.Sh "Tk Generic Methods"
.IX Subsection "Tk Generic Methods"
.IP "\(bu" 4
Tk::Font
.IP "\(bu" 4
Tk::send
.IP "\(bu" 4
Tk::tkvars
.IP "\(bu" 4
Tk::Widget
.IP "\(bu" 4
Tk::X11Font
.Sh "User Interaction"
.IX Subsection "User Interaction"
.IP "\(bu" 4
Tk::DropSite
.IP "\(bu" 4
Tk::Clipboard
.IP "\(bu" 4
Tk::focus
.IP "\(bu" 4
Tk::grab
.IP "\(bu" 4
Tk::selection
.Sh "Creating and Configuring Widgets"
.IX Subsection "Creating and Configuring Widgets"
.IP "\(bu" 4
Tk::CmdLine
.IP "\(bu" 4
Tk::MainWindow
.IP "\(bu" 4
Tk::option
.IP "\(bu" 4
Tk::options
.IP "\(bu" 4
Tk::palette
.IP "\(bu" 4
Tk::Xrm
.Sh "Popups and Dialogs"
.IX Subsection "Popups and Dialogs"
.IP "\(bu" 4
Tk::chooseColor
.IP "\(bu" 4
Tk::ColorEditor
.IP "\(bu" 4
Tk::Dialog
.IP "\(bu" 4
Tk::Dialog
.IP "\(bu" 4
Tk::FileSelect
.IP "\(bu" 4
Tk::getOpenFile
.IP "\(bu" 4
Tk::messageBox
.Sh "Derived Widgets"
.IX Subsection "Derived Widgets"
.IP "\(bu" 4
Tk::composite
.IP "\(bu" 4
Tk::configspec
.IP "\(bu" 4
Tk::Derived
.IP "\(bu" 4
Tk::mega
.IP "\(bu" 4
Tk::ROText
.IP "\(bu" 4
Tk::Scrolled
.IP "\(bu" 4
Tk::TextUndo
.IP "\(bu" 4
Tk::Reindex
.IP "\(bu" 4
Tk::Pane
.IP "\(bu" 4
Tk::ProgressBar
.Sh "C Programming"
.IX Subsection "C Programming"
.IP "\(bu" 4
Internals
.IP "\(bu" 4
pTk
.IP "\(bu" 4
3DBorder
.IP "\(bu" 4
BackgdErr
.IP "\(bu" 4
BindTable
.IP "\(bu" 4
CanvPsY
.IP "\(bu" 4
CanvTkwin
.IP "\(bu" 4
CanvTxtInfo
.IP "\(bu" 4
Clipboard
.IP "\(bu" 4
ClrSelect
.IP "\(bu" 4
ConfigWidg
.IP "\(bu" 4
ConfigWind
.IP "\(bu" 4
CoordToWin
.IP "\(bu" 4
CrtErrHdlr
.IP "\(bu" 4
CrtGenHdlr
.IP "\(bu" 4
CrtImgType
.IP "\(bu" 4
CrtItemType
.IP "\(bu" 4
CrtMainWin
.IP "\(bu" 4
CrtPhImgFmt
.IP "\(bu" 4
CrtSelHdlr
.IP "\(bu" 4
CrtWindow
.IP "\(bu" 4
DeleteImg
.IP "\(bu" 4
DoOneEvent
.IP "\(bu" 4
DoWhenIdle
.IP "\(bu" 4
DrawFocHlt
.IP "\(bu" 4
EventHndlr
.IP "\(bu" 4
EventInit
.IP "\(bu" 4
FileHndlr
.IP "\(bu" 4
FindPhoto
.IP "\(bu" 4
FontId
.IP "\(bu" 4
FreeXId
.IP "\(bu" 4
GeomReq
.IP "\(bu" 4
GetAnchor
.IP "\(bu" 4
GetBitmap
.IP "\(bu" 4
GetCapStyl
.IP "\(bu" 4
GetClrmap
.IP "\(bu" 4
GetColor
.IP "\(bu" 4
GetCursor
.IP "\(bu" 4
GetFont
.IP "\(bu" 4
GetFontStr
.IP "\(bu" 4
GetGC
.IP "\(bu" 4
GetImage
.IP "\(bu" 4
GetJoinStl
.IP "\(bu" 4
GetJustify
.IP "\(bu" 4
GetOption
.IP "\(bu" 4
GetPixels
.IP "\(bu" 4
GetPixmap
.IP "\(bu" 4
GetRelief
.IP "\(bu" 4
GetRootCrd
.IP "\(bu" 4
GetScroll
.IP "\(bu" 4
GetSelect
.IP "\(bu" 4
GetUid
.IP "\(bu" 4
GetVisual
.IP "\(bu" 4
GetVRoot
.IP "\(bu" 4
HandleEvent
.IP "\(bu" 4
IdToWindow
.IP "\(bu" 4
ImgChanged
.IP "\(bu" 4
InternAtom
.IP "\(bu" 4
MainLoop
.IP "\(bu" 4
MaintGeom
.IP "\(bu" 4
MainWin
.IP "\(bu" 4
ManageGeom
.IP "\(bu" 4
MapWindow
.IP "\(bu" 4
MeasureChar
.IP "\(bu" 4
MoveToplev
.IP "\(bu" 4
Name
.IP "\(bu" 4
NameOfImg
.IP "\(bu" 4
OwnSelect
.IP "\(bu" 4
ParseArgv
.IP "\(bu" 4
Preserve
.IP "\(bu" 4
QWinEvent
.IP "\(bu" 4
Restack
.IP "\(bu" 4
RestrictEv
.IP "\(bu" 4
SetAppName
.IP "\(bu" 4
SetClass
.IP "\(bu" 4
SetGrid
.IP "\(bu" 4
SetVisual
.IP "\(bu" 4
Sleep
.IP "\(bu" 4
StrictMotif
.IP "\(bu" 4
TextLayout
.IP "\(bu" 4
TimerHndlr
.IP "\(bu" 4
Tk_Init
.IP "\(bu" 4
WindowId
.Sh "Implementation"
.IX Subsection "Implementation"
.IP "\(bu" 4
Tk::Eventloop
.IP "\(bu" 4
Tk::Item
.IP "\(bu" 4
Tk::Submethods
.IP "\(bu" 4
Tk::WidgetDemo
.IP "\(bu" 4
Tk::widgets
.Sh "Experimental Modules"
.IX Subsection "Experimental Modules"
.IP "\(bu" 4
Tk::Common
.IP "\(bu" 4
Tk::SunConst
.IP "\(bu" 4
Tk::WinPhoto
.Sh "Other Modules and Languages"
.IX Subsection "Other Modules and Languages"
.IP "\(bu" 4
Tk::Compile
.IP "\(bu" 4
Tk::Tcl\-perl
.IP "\(bu" 4
Tk::X
.SH "AUTHOR"
.IX Header "AUTHOR"
Nick Ing-Simmons
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIperl\fR\|(1), \fIwish\fR\|(1).


OST2.design/tools/perl-5.8.0/lib/site_perl/5.8.0/CommandTerm/Tk.pm
package CommandTerm::Tk;

use strict;
use Tk;
use Tk::LabEntry;
use Term::ReadLine;
use FileHandle;
use POSIX;
use CommandTerm;

@CommandTerm::Tk::ISA = qw(CommandTerm Tk::Frame);
Tk::Widget->Construct('CommandTermTk');

use vars qw(
  %DerivedConfig
);

BEGIN {
  %DerivedConfig = (
     'width'               => '80',
     'height'              => '24',
     'textFont'            => '-*-courier-medium-r-*-*-12-*-*-*-*-*-*-*',
     'entryFont'           => '-*-courier-medium-r-*-*-12-*-*-*-*-*-*-*',
     'promptFont'          => '-*-courier-bold-r-*-*-12-*-*-*-*-*-*-*',
     'saveTextLines'       => '5000',
     'dataFace'            => 'normal',
     'statusFace'          => 'italic',
     'errorFace'           => 'bold',
  );
}

sub Populate {
  my ($self, $args) = @_;

  my $name        = delete $args->{'-name'};
  my $toplevel    = delete $args->{'-toplevel'};
  my $cfgfile     = delete $args->{'-cfgfile'};
  my $app_config  = delete $args->{'-app_config'};

     $self->InitSignalHandlers;
     $self->Init($name, $cfgfile, $app_config, \%DerivedConfig);
     $self->NoPipes(1);

  my $prompt      = delete $args->{'-prompt'} || $self->GetPrompt;

  my $width                       = $self->{'Config'}{'width'};
  my $height                      = $self->{'Config'}{'height'};
  my $textFont                    = $self->{'Config'}{'textFont'};
  my $entryFont                   = $self->{'Config'}{'entryFont'};
  my $promptFont                  = $self->{'Config'}{'promptFont'};
  my $italicFont                  = '-*-courier-medium-o-*-*-12-*-*-*-*-*-*-*';

     $self->{'TopLevel'}          = $toplevel;
     $self->{'RealWidth'}         = $width;
     $self->{'SaveTextLines'}     = $self->{'Config'}{'saveTextLines'};
     $self->{'DataFace'}          = $self->{'Config'}{'dataFace'};
     $self->{'StatusFace'}        = $self->{'Config'}{'statusFace'};
     $self->{'ErrorFace'}         = $self->{'Config'}{'errorFace'};

     $self->SUPER::Populate($args);

  my $TextFrame = $self->Frame;
  my $Text = $TextFrame->Text( 
                      -relief  => 'sunken',
                      -wrap    => 'word',
                      -state   => 'disabled',
                      -font    => $textFont, 
                      -width   => $width,
                      -height  => $height,
                   );
     $Text->bind( "<Configure>"  => sub { $self->FigureOutWidth } );
     $Text->tagConfigure( $self->DataFace,   -font => $textFont );
     $Text->tagConfigure( $self->StatusFace, -font => $italicFont );
     $Text->tagConfigure( $self->ErrorFace,  -font => $promptFont );

  my $ScrollBar = $self->Scrollbar( -command => [$Text => 'yview']);
     $Text->configure( -yscrollcommand => [$ScrollBar => 'set'] );
  my %nopad = ( -pady => 0, -ipady => 0, -padx => 0, -ipadx => 0 );
     $ScrollBar->pack( %nopad, -side => 'right', -fill => 'y' );
     $Text->pack( %nopad, -side => 'left', -expand => 'both', -fill => 'both' );

  my $EntryFrame = $self->Frame;
  my $Entry = $EntryFrame->Component(
                              LabEntry      => 'labEntry',
                              -label        => $prompt,
                              -labelPack    => [ -side => 'left' ],
                              -relief       => 'sunken',
                              -font         => $entryFont,
                              );
     $Entry->Subwidget('label')->configure( -font => $promptFont );
     $Entry->pack( -fill => 'x' );
     $Entry->bind( '<Delete>'     => 'Backspace' );
     $Entry->bind( '<Return>'     => sub { $self->ProcessEntry; } );
     $Entry->bind( '<Tab>'        => sub { $self->Complete; } );
     $Entry->bind( '<Control-t>'  => sub { $self->TogglePrompt; } );
     $Entry->bind( '<Up>'         => sub { $self->UpHistory; } );
     $Entry->bind( '<Control-p>'  => sub { $self->UpHistory; } );
     $Entry->bind( '<Down>'       => sub { $self->DownHistory; } );
     $Entry->bind( '<Control-n>'  => sub { $self->DownHistory; } );
     $Entry->bind( '<Control-l>'  => sub { $self->clear; } );
     $Entry->bind( '<Control-u>'  => sub { $Entry->delete(0, 'end'); } );
     $Entry->bind( '<Control-c>'  => sub { $self->HandleInterrupt; } );
     $Entry->bind( '<Home>'       => sub { $self->Home; } );
     $Entry->bind( '<End>'        => sub { $self->End; } );
     $Entry->bind( '<Prior>'      => sub { $self->ScrollUp; } );
     $Entry->bind( '<Next>'       => sub { $self->ScrollDown; } );
     $Entry->Subwidget('entry')->bindtags(
                   [$Entry->Subwidget('entry'),
                   ref($Entry->Subwidget('entry')), 'all']
     );

     # Button 2 release events will paste the X selection into the 
     # entry widget no matter where the cursor is over $self
     $self->bind( 'Tk::Entry', '<ButtonRelease-2>', undef);
     $self->bind( 'Tk::Text',  '<ButtonRelease-2>', undef);
     $self->bind( 'all',       '<ButtonRelease-2>' =>
        sub { 
          Tk::catch {
            $Entry->insert($Entry->index('insert'), $self->SelectionGet)
          };
          $Entry->SeeInsert;
        }
     );

     $EntryFrame->pack( -side => 'bottom' , -fill => 'x' );
     $TextFrame->pack( -side => 'top' , -fill => 'both', -expand => 'both' );

     $self->Advertise('Text'  => $Text );
     $self->Advertise('Entry' => $Entry );

     $self->InitReadLine;

     $Entry->Subwidget('entry')->focus;
}

sub history {
  my ($self, @args) = @_;
  my ($opt_h, $opt_r, $opt_n);
  
  $self->get_options(\@args, ['no_pass_through'], 'h' => \$opt_h, 'r' => \$opt_r) or return;

  $opt_n = shift @args;

  if( $opt_n and $opt_n !~ /^\d+$/ ) {
    $self->PrintError("Argument to history '$opt_n' isn't numeric.\n");
    return;
  }

  my $str = '';
  my @history = $self->ReadLine->GetHistory;
  if( $opt_r ) {
    for(my $i=$#history; $i>=0; $i--) {
      $str .= ($opt_h ? '' : "\t$i\t" . "$history[$i]\n");
      last if $opt_n and ($#history - $i + 1) >= $opt_n;
    }
  } else {
    $opt_n = $#history unless defined($opt_n);
    for(my $i=$#history-$opt_n+1; $i<=$#history; $i++) {
      $str .= ($opt_h ? '' : "\t$i\t" . "$history[$i]\n");
    }
  }
  $self->PrintStatus($str);
}

sub ShellExec {
  my ($self, $cmd, @args) = @_;

  if( $cmd eq 'cd' ) {
    my ($dir) = @args;

    $dir = '' unless $dir;
   
    $dir = $self->GlobExpand($dir);

    return unless defined($dir);

    $dir =~ s/^"(.*)"$/$1/;
    $dir =~ s/^'(.*)'$/$1/;

    if( chdir $dir ) {
      $self->PrintData( getcwd . "\n" );
    } else {
      $self->PrintError("Couldn't cd to '$dir': $!\n");
    }
  } else {
    if( $self->Which($cmd) ) {
      $self->PrintData( `$cmd @args` );
    } else {
      $self->PrintError("Command not found: '$cmd'\n");
    }
  }
}

sub ProcessEntry {
  my ($self) = @_;
  my $Entry    = $self->Subwidget('Entry');
  my $line     = $Entry->get;

  chomp($line);

  $Entry->delete(0, 'end');

  $self->ProcessLine($line);
}

sub ProcessLine {
  my ($self, $line) = @_;
  my $ReadLine = $self->ReadLine;

  $self->CloseRedirection;

  $self->PrintError("> $line\n");

  $self->SUPER::ProcessLine($line);

  $self->Subwidget('Entry')->configure( -label => $self->GetPrompt );

  # Since we're not using ReadLine interactively, set the history position
  # to the end every time through.
  $ReadLine->history_set_pos( scalar($ReadLine->GetHistory()) );
}

sub InitReadLine {
  my ($self) = @_;

  $self->SUPER::InitReadLine;

  my $NULLOUT = FileHandle->new('> /dev/null') or 
         die "Couldn't open '/dev/null': $!\n";

  $self->Attribs->{'outstream'} = $NULLOUT;

  #In Tk mode, don't want any dinging at all, it's alarming. :-)
  $self->ReadLine->parse_and_bind('set bell-style none');
}

sub TogglePrompt {
  my ($self) = @_;

  &{ $self->TogglePromptCallback };
  
  $self->Subwidget('Entry')->configure( -label => $self->GetPrompt );

  Tk->break;
} 
  
sub Complete {
  my ($self) = @_;
  my $ReadLine = $self->ReadLine;
  my $Attribs  = $self->Attribs;
  my $Entry    = $self->Subwidget('Entry');
  my $line     = $Entry->get;

  $Attribs->{'line_buffer'} = $line;
  $Attribs->{'point'}       = $Entry->index('insert');
  $Attribs->{'end'}         = length($Attribs->{'line_buffer'});

  $ReadLine->complete_internal(ord("\t"));

  if( $line ne $Attribs->{'line_buffer'} ) {
    # completion took place
    $Entry->delete(0, 'end');
    $Entry->insert(0, $Attribs->{'line_buffer'});
    $Entry->icursor($Attribs->{'point'});
    $Entry->SeeInsert;
  } else {
    # no completion took place, list possible completions
    $ReadLine->complete_internal(ord('?'));
  }

  Tk->break;
}

sub PrintCompletions {
  my ($self, $matchlist_ref, $num_matches, $longest, $striplen) = @_;

  my $match;
  my @matches = ();
  my $appended = 0;
  for(my $i=1; $i<=$num_matches; $i++) {
    $match = $matchlist_ref->[$i];
    if( -d $match ) {
      $match .= '/';
      $appended = 1;
    } elsif( -l $match ) {
      $match .= '@';
      $appended = 1;
    } elsif( -x $match ) {
      $match .= '*';
      $appended = 1;
    }
    push @matches, substr $match, $striplen;
  }
  $longest++ if $appended;

  my $width = $self->RealWidth;
  my $num_cols = int($width/($longest+1)) || 1;
  my $skip = int($num_matches/$num_cols);
  $skip++ if ($num_matches/$num_cols) != $skip;

  my ($i, $j, $index);
  my $str = '';
  for($i=0; $i<$skip; $i++) {
    for($j=0; $j<$num_cols; $j++) {
      $index = $i + ($j * $skip);
      if( $index < $num_matches ) {
        $str .= sprintf("%-${longest}s ", $matches[$index]);
      }
    }
    $str =~ s/ $/\n/;
  }
  $self->PrintStatus($str);
  $self->PrintError(">\n");
}

sub Shutdown { }

sub Quit {
  Tk::exit;
}

sub Title {
  my ($self, $titlestr) = @_;

  $self->TopLevel->title($titlestr) if $self->TopLevel;
}

sub Icon {
  my ($self, $iconstr) = @_;

  $self->TopLevel->iconname($iconstr) if $self->TopLevel;
}

sub Print {
  my ($self, $how, @what) = @_;
  my $Text = $self->Subwidget('Text');
  my ($top, $bottom) = $Text->yview;
  my $LogFH = $self->LogFH;
  my $ReDirFH = $self->ReDirFH;

  my $what = join '', @what;
  print $LogFH $what if defined $LogFH;

  if( $ReDirFH ) {
    print $ReDirFH $what;
  } else {

    $Text->configure( -state => 'normal' );
    if( $what =~ /[\b]/ ) {
      foreach my $bsline (split("(\b)", $what)) {
        if( $bsline eq "\b" ) {
          $Text->delete('end -2 chars');
        } else {
          $Text->insert('end', $bsline, $how );
        }
      }
    } else {
     $Text->insert('end', $what, $how );
    }

    my $deletelines = ($Text->index('end') - 2.0) - $self->SaveTextLines;
    $Text->delete( '1.0', $deletelines . ".0" ) if $deletelines > 1;

    $Text->see( 'end' ) if $bottom == 1.0;

    $Text->configure( -state => 'disabled' );

    $self->idletasks;
  }
}

sub UpHistory {
  my ($self) = @_;
  my $Entry = $self->Subwidget('Entry');

  my $line = $self->ReadLine->previous_history;

  if( $line ne '' ) {
    $Entry->delete(0, 'end'); 
    $Entry->insert(0, $line );
  }
}

sub DownHistory {
  my ($self) = @_;
  my $Entry = $self->Subwidget('Entry');

  my $line = $self->ReadLine->next_history;

  $Entry->delete(0, 'end');
  $Entry->insert(0, $line );
} 

sub HandleInterrupt {
  my ($self) = @_;
  my $Entry = $self->Subwidget('Entry');

  if( $Entry->get ne '' ) {
    $Entry->delete(0, 'end');
  } else {
    &{ $self->HandleInterruptCallback };
  }
}

sub DefaultInterruptCallback {
  my ($self) = @_;

  $self->PrintError(">\n");
  $self->UpdatePrompt(1);
}

sub FigureOutWidth {
  my ($self) = @_;
  my ($i, $xleft, $xright, $ytop, $ybot, $line);
  my $Text = $self->Subwidget('Text');

  ($ytop, $ybot) = $Text->yview;
  $line = $Text->index( 'end' ) - 1;
  $Text->configure( -state => 'normal', -wrap => 'none' );
  $Text->see( 'end' );
  $i = 0;
  ($xleft, $xright) = $Text->xview;
  while( $xleft == 0.0 and $xright == 1.0 ) {
    $Text->insert( 'end' => 'O', 'italic' );
    ($xleft, $xright) = $Text->xview;
    $i++;
  }
  $Text->delete( "$line.0", 'end -1 chars' );
  $Text->yview( 'moveto' => $ytop );
  $Text->configure( -state => 'disabled', -wrap => 'word' );
  $self->RealWidth($i--);
}

sub clear {
 my ($self) = @_;
 my $Text = $self->Subwidget('Text');

 $Text->configure( -state => 'normal' );
 $Text->delete( '1.0', 'end' );
 $Text->see( '1.0' );
 $Text->configure( -state => 'disabled' );
}

sub Home {
  my ($self) = @_;
  my $Text = $self->Subwidget('Text');

  $Text->configure( -state => 'normal' );
  $Text->see( '1.0' );
  $Text->configure( -state => 'disabled' );
}

sub End {
  my ($self) = @_;
  my $Text = $self->Subwidget('Text');

  $Text->configure( -state => 'normal' );
  $Text->see( 'end' );
  $Text->configure( -state => 'disabled' );
}

sub ScrollUp {
  my ($self) = @_;
  my $Text = $self->Subwidget('Text');

  $Text->configure( -state => 'normal' );
  $Text->yview( 'scroll' => -1, 'pages');
  $Text->configure( -state => 'disabled' );
}

sub ScrollDown {
  my ($self) = @_;
  my $Text = $self->Subwidget('Text');

  $Text->configure( -state => 'normal' );
  $Text->yview( 'scroll' => 1, 'pages');
  $Text->configure( -state => 'disabled' );
}

1;
